SIPS/sip-7.md

```diff
@@ -2,307 +2,205 @@
+### Literal
 
-interface SpacerComponent extends BaseComponent {
-  type: 'spacer';
+```typescript
+interface Literal extends Node {
+  value: any;
 }
-
-interface SpinnerComponent extends BaseComponent {
-  type: 'spinner';
-}
-
-type Component =
-  | PanelComponent
-  | HeadingComponent
-  | TextComponent
-  | ButtonComponent
-  | DividerComponent
-  | SpacerComponent
-  | SpinnerComponent;
 ```
 
-### JSON-RPC methods
-
-#### `snap_showInterface`
-
-The `snap_showInterface` method lets Snaps show a user interface arbitrarily. The params MUST be an object, with the following properties:
+## Core Components
 
-- `ui` (`Component`): The custom UI to show.
+> Defining a complete set of components used in the wallet is outside of scope for this SIP.
+>
+> Only a core set of components is defined, delegating definition of a complete set of useful components to other SIPs.
 
-It MUST return a `string`, which is used as identifier for the user interface. This identifier MUST be unique for each invocation of this method.
+### UI Root
 
-```tsx
-interface SnapShowInterfaceRequest {
-  method: 'snap_showInterface';
-  params: {
-    ui: Component;
-  };
-}
-
-type SnapShowInterfaceResponse = string;
+```typescript
+type UIRoot = Panel | Panel[];
 ```
 
-#### `snap_updateInterface`
+- `UIRoot` - The root of the custom UI hierarchy. A UI hierarchy MUST be always nested inside a flat list of Panels.
 
-The `snap_updateInterface` method lets Snaps update an existing user interface. The params MUST be an object, with the following properties:
+### Panel
 
-- `id` (`string`): The ID of the custom UI to update.
-- `ui` (`Component`): The new custom UI to show. This will replace the existing UI.
-
-This method does not return anything.
-
-```tsx
-interface SnapUpdateInterfaceRequest {
-  method: 'snap_updateInterface';
-  params: {
-    id: string;
-    ui: Component;
-  };
+```typescript
+interface Panel extends Parent {
+  type: "panel";
+  children: Node[];
 }
 ```
 
-#### `snap_resolveInterface`
+- `Panel` - A block of visually grouped components.
 
-Snaps that require user input in order to continue execution, i.e., checking if a user pressed the confirm or deny button, can use the `snap_resolveInterface` method to provide the resolved value. This is primarily intended to be used in the `onUserInput` entry point, see the section “User input” below.
+## Permissions
 
-The params MUST be an object, with the following properties:
+## Communication with the wallet
 
-- `id` (`string`): The ID of the custom UI to resolve.
-- `value` (`Json`): A JSON-compatible value to resolve the interface with. This value is passed to the `snap_readInterface` call.
+A snap intending to use the custom UI SHOULD expose a `ui` object using CommonJS.
 
-This method does not return anything.
-
-```tsx
-interface SnapResolveInterfaceRequest {
-  method: 'snap_resolveInterface';
-  params: {
-    id: string;
-    value: Json;
-  };
-}
+```typescript
+module.exports.ui = new MyUI();
```

Comment:
I don't understand the need for a class here. This could be implemented using an object, as the class likely won't hold any meaningful state.

Reply:
No need for a class. It's to showcase that it should be an *instance* or an object, rather than prototype of a class.

Same explanation happens in SIP-2